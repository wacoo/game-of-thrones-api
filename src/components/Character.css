.cell-detail {
  height: 200px;
  display: grid;
  grid-template-columns: 1fr auto;
  justify-content: space-between;
  align-items: center;
  padding: 12px;
  color: #fff;
}

.cell-detail:nth-child(even) {
  background-color: #004d00;
}

.cell-detail:nth-child(odd) {
  background-color: #022b02;
}

.cell-detail h2 {
  width: 150px;
}

.data {
  width: 150px;
  font-family: Arial, Helvetica, sans-serif;
  font-weight: 500;
  font-size: 14px;
}

.character-sub {
  width: 100%;
}

.character-main img {
  width: 150px;
}

.character-main {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #267c26;
}

.character-main i {
  margin: 12px;
  font-size: 30px;
  color: #fff;
}

.character-main div {
  display: flex;
  width: 100%;
  height: 150px;
  background-color: #267c26;
  margin: 0 12px;
  align-items: center;
  justify-content: space-around;
}

.character-main h2 {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 25px;
  color: #fff;
}
