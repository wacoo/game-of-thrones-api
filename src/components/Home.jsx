import './Home.css';
import { useDispatch } from 'react-redux';
import Books from './Books';
import { getFiltered } from '../redux/books/booksSlice';

const Home = () => {
  const dispatch = useDispatch();

  const filterCountry = (value) => {
    dispatch(getFiltered(value));
  };

  return (
    <div className="container">
      <div className="main-area">
        <input
          type="text"
          placeholder="Search..."
          onChange={(e) => filterCountry(e.target.value)}
        />
      </div>
      <div className="sub-area">
        <Books />
      </div>
    </div>
  );
};

export default Home;
