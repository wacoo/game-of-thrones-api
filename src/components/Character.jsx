import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import './Character.css';
import { v4 as uuidv4 } from 'uuid';

const Character = () => {
  const dispatch = useDispatch();
  const selectedBook = useSelector(
    (state) => state.books.selectedBook,
  );

  const characters = selectedBook.characterData;

  useEffect(() => {
    // Add any necessary logic or side effects related to the selectedBook here
  }, [selectedBook]);

  return (
    <div className="container">
      <div className="character-main">
        <a href="/">
          <i className="fa fa-arrow-circle-left" />
        </a>
        <div>
          <h2>{selectedBook.name}</h2>
          {/* <img src={selectedCountry.map} alt="" /> */}
        </div>
      </div>
      <div className="character-sub">
        {characters && characters.length > 0 ? (
          characters.map((character) => {
            const uuid = uuidv4();
            const {
              name, gender, born, died, aliases, titles
            } = character;
            return (
              <div className="cell-detail" key={uuid}>
                <h2>{name}</h2>
                <div className="data">
                  <span>{`Gender: ${gender}`}</span>
                  <br />
                  <span>{`Born: ${born}`}</span>
                  <br />
                  <span>{`Died: ${died}`}</span>
                  <br />
                  <span>{`Alias: ${aliases[0] ? aliases[0] : ''}`}</span>
                  <br />
                  <span>{`Title: ${titles[0]}`}</span>
                </div>
              </div>
            );
          })
        ) : (
          <p>No character data available</p>
        )}
      </div>
    </div>
  );
};

export default Character;