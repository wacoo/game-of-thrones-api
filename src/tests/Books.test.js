import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { render, screen } from '@testing-library/react';
import Home from '../components/Home';
import Character from '../components/Character';
import '@testing-library/jest-dom/extend-expect';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

const fakeData = {
  books: [
    {
      name: 'A Game of Thrones',
      authors: ['George R.R Martin'],
      characterData: [
        {          
          name: 'Walder',
          gender: 'Male'
        }
      ]
    },
    {
      name: 'A Clash of Kings',
      authors: ['George R.R Martin'],
      characterData: [
        {          
          name: 'Walder',
          gender: 'Male'
        }
      ]
    },
  ],
};

describe('Test components render', () => {
  it('Tests Home', () => {
    useDispatch();
    useSelector.mockReturnValue(fakeData.books);
    const homeComponent = renderer.create(
      <BrowserRouter>
        <Home />
      </BrowserRouter>,
    );
    expect(homeComponent).toMatchSnapshot();
  });

  it('Tests Character', () => {
    useDispatch();
    useSelector.mockReturnValue(fakeData.books[0].characterData[0]);
    const homeComponent = renderer.create(
      <BrowserRouter>
        <Character />
      </BrowserRouter>,
    );
    expect(homeComponent).toMatchSnapshot();
  });
});

describe('Test existance of elements in components', () => {
  it('Tests existance of the button "A Clash of Kings" in Home ', () => {
    useDispatch();
    useSelector.mockReturnValue(fakeData.books);
    render(<Home />);
    const h2 = screen.getByText('A Clash of Kings');
    expect(h2).toBeInTheDocument();
  });
  it('Tests existance of "Walder" entry in Characters page ', () => {
    useDispatch();
    useSelector.mockReturnValue(fakeData.books[0].characterData[0]);
    render(<Character />);
    const h2 = screen.getByText('Walder');
    expect(h2).toBeInTheDocument();
  });
});
