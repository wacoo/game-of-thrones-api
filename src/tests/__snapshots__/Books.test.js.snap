// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test components render Tests Character 1`] = `
<div
  className="container"
>
  <div
    className="character-main"
  >
    <a
      href="/"
    >
      <i
        className="fa fa-arrow-circle-left"
      />
    </a>
    <div>
      <h2>
        Walder
      </h2>
    </div>
  </div>
  <div
    className="character-sub"
  >
    <p>
      No character data available
    </p>
  </div>
</div>
`;

exports[`Test components render Tests Home 1`] = `
<div
  className="container"
>
  <div
    className="main-area"
  >
    <input
      onChange={[Function]}
      placeholder="Search..."
      type="text"
    />
  </div>
  <div
    className="sub-area"
  >
    <button
      className="cell"
      onClick={[Function]}
      style={
        Object {
          "backgroundRepeat": "no-repeat",
          "backgroundSize": "contain",
        }
      }
      type="button"
    >
      <div
        className="inside-icon-wrapper"
      >
        <i
          className="fa fa-arrow-circle-right"
        />
      </div>
      <div
        className=""
      >
        <h2>
          A Game of Thrones
        </h2>
        <div
          className="books"
        >
          Author: George R.R Martin
          <br />
          ISBN: undefined
          <br />
          No. of Page: undefined
          <br />
          Publusher: undefined
          <br />
          Release date: undefined
        </div>
      </div>
    </button>
    <button
      className="cell"
      onClick={[Function]}
      style={
        Object {
          "backgroundRepeat": "no-repeat",
          "backgroundSize": "contain",
        }
      }
      type="button"
    >
      <div
        className="inside-icon-wrapper"
      >
        <i
          className="fa fa-arrow-circle-right"
        />
      </div>
      <div
        className=""
      >
        <h2>
          A Clash of Kings
        </h2>
        <div
          className="books"
        >
          Author: George R.R Martin
          <br />
          ISBN: undefined
          <br />
          No. of Page: undefined
          <br />
          Publusher: undefined
          <br />
          Release date: undefined
        </div>
      </div>
    </button>
  </div>
</div>
`;
